src/v/archival/archival_policy.cc

```diff
@@ -75,6 +90,71 @@ bool archival_policy::upload_deadline_reached() {
     return _upload_deadline < now;
 }
 
+archival_policy::lookup_result archival_policy::find_compacted_segment(
+  model::offset last_compacted_offset, storage::log log) {
+    vlog(
+      archival_log.debug,
+      "Finding next compacted segment for {} , last compacted offset: {}",
+      _ntp,
+      last_compacted_offset);
+
+    auto maybe_plog = get_concrete_log_impl(std::move(log));
+    if (!plog) {
+        vlog(
+          archival_log.debug,
+          "Finding next compacted segment for {}: can't find candidate, no "
+          "segments or "
+          "in-memory log",
+          _ntp);
+        return {};
+    }
+
+    auto plog = maybe_plog.value();
+    const auto& set = plog->segments();
+    auto it = set.lower_bound(last_compacted_offset);
+
+    // iterator can reach end for two reasons, either we did not find a segment
+    // higher than last_compacted_offset OR last_compacted_offset lies in a gap
+    // where there are no segments.
```

Comment:
I think that the comment is a bit misleading. It's not possible to have a gap in the log but the log could be prefix-truncated. In this case we just need to move `last_compacted_offset` a bit forward.

Reply:
Thanks for clearing it up, I will fix the comment.

Reply:
I have updated the comments, I also renamed the input from `last_compacted_offset` to `start_offset` as last offset does not make sense, we are finding the segment containing the offset so the start offset will be last compacted offset + 1.